src/apps/workspaces/crud/user_applet_access.py

```diff
@@ -532,24 +568,45 @@ async def get_workspace_respondents(
         if query_params.search:
             query = query.having(_WorkspaceRespondentSearch().get_clauses(query_params.search))
 
+        # Get total count before ordering and pagination to evaluate eligibility for manual sorting
         coro_total = self._execute(select(count()).select_from(query.with_only_columns(UserSchema.id).subquery()))
+        total = (await coro_total).scalar()
+
+        ordering = _WorkspaceRespondentOrdering()
+        manual_order_fields = None
 
         if query_params.ordering:
-            query = query.order_by(*_WorkspaceRespondentOrdering().get_clauses(*query_params.ordering))
-        query = paging(query, query_params.page, query_params.limit)
-        coro_data = self._execute(query)
-        res_data, res_total = await asyncio.gather(coro_data, coro_total)
-        data = parse_obj_as(list[WorkspaceRespondent], res_data.all())
-        total = res_total.scalar()
-        return data, total
+            if total < RESPONDENTS_MANUAL_SORT_LIMIT:
+                manual_order_fields = ordering.get_manual_fields(*query_params.ordering)
+            # Only perform SQL ORDER BY if all requested ordering fields can be done by SQL
+            if not manual_order_fields:
+                query = query.order_by(*ordering.get_clauses(*query_params.ordering))
+
+        # If able to use SQL ordering, also use SQL-based paging; else paginate post-execute
+        if not manual_order_fields:
+            query = paging(query, query_params.page, query_params.limit)
+
+        res_data = await self._execute(query)
+
+        data = res_data.all()
+
+        # If sorting manually, both sort and paginate manually
+        if manual_order_fields:
+            data = ordering.manual_sort(data, manual_order_fields)
+            data = ordering.manual_paginate(data, query_params.page, query_params.limit)
+
+        data = parse_obj_as(list[WorkspaceRespondent], data)
+        ordering_fields = ordering.get_ordering_fields(total < RESPONDENTS_MANUAL_SORT_LIMIT)
+
+        return data, total, ordering_fields
```

Comment:
I'm not sure I follow 100% what you're getting at, but I'll try to offer an explanation as best I can.

The sortable fields (`ordering_fields`) is now _dynamic_ for `respondents` and `managers` endpoints, as the list of fields that can be ordered by is conditional on the # of records. When performing manual sorting (to support sorting by encrypted fields), we have to place a limit on the # of records we can sort by (via the `RESPONDENTS_MANUAL_SORT_LIMIT` and `MANAGERS_MANUAL_SORT_LIMIT` constants), so as to limit resource use since sorting manually uses extra memory in Python. When # of records exceeds this limit, we can't do manual sorting, which means we can no longer sort by encrypted fields – `nicknames` in the case of respondents, and `first_name`, `last_name`, and `email` in the case of Managers. Therefore, the list of sortable fields, expressed by `ordering_fields`, can change.

The admin panel needs to be informed of which columns can be sorted by so that it can disable sorting buttons on those columns as required, and avoid giving the user the impression that they can sort by a column where sorting actually doesn't work. This has been implemented in https://github.com/ChildMindInstitute/mindlogger-admin/pull/1767.

Please let me know if I've answered your question, or if I've misunderstood.

Reply:
@vshvechko note that product is aligned with this approach and the UX impact it has. Now sorting with pagination works as long as it's under the threshold limit.

Reply:
@farmerpaul Thanks a lot for the explanation, it's very helpful. My initial concern was about the appropriate place for manual sorting, as it seems more related to business logic rather than repository responsibility. I don't have strong vision what is the right place for it, just wanted to open the discussion.

We might want to consider using general constants for manual sorting since it could be applicable for other purposes in the future, not just for respondents and managers.

Additionally, I want to mention that we have SQL functions like `decrypt_internal` for decrypting fields at the database level. This might not be suitable for your case, as the dataset can be large, and decryption at the database level would definitely slow down query execution. Just wanted to let you know about this option.
Usage
`decrypt_internal(field, decode('<encryption_key_hex>', 'hex'))`

Reply:
🤯 Thanks for sharing that SQL function with me, @vshvechko. I suspect using that function will be a lot cleaner than the approach I used. I'm not sure how doing the decryption SQL-side would be less performant than doing it Python-side – in both cases decryption has to run on all records, so I think having the record limits would still be worthwhile.

I agree making the constants more general might be in order. I used 2 separate constants at first because I didn't want to assume that memory needs to manually sort respondents & managers would be the same. (In the case of managers, it seems unlikely that the limit would ever be approached, anyway.)

I'll try the approach using `decrypt_internal`, which hopefully would make the discussion around where to place manual sorting logic not longer pertinent.

Reply:
Normally I avoid use database resource more than need, it usually more expensive than normal servers. But, In this case we already have a limit and using decrypt_internal will make code cleaner, so I aggred give a try using `decrypt_internal`